pkg/util/error_list.go

```diff
@@ -45,10 +45,67 @@ func (list errorList) Error() string {
 	return result
 }
 
+// Except returns any errors (or nested errors, if the list contains
+// nested []error slices) for which all fns return false. If no errors
+// remain a nil list is returned. The resulting list will have all
+// nested []error slices flattened as a side effect.
```

Comment:
Actually not, it's from multiple sources - so if I have:

* directory 1
* directory 2
* stdin
* selector on pods

Each of those is a source of runtime.Objects.  The first directory could hit an error on file dir1/foo.json, and stdin could hit an error on the 3rd object in the stream (unparseable, or no kind), and the pod selector could hit an error talking to the server.  So the construct is:

```
[]error{
  []error{
    "file dir1/foo.json permission denied"
    "file dir1/bar.json permission denied"
  },
  "error reading stdin",
  []error{
    "unable to read from server X"
  },
}
```

Each error is individual (a file, entry in stream, or potentially entry in a list from server that is malformed for some reason), and the whole set is filtered.

The error from the visitor is returned as "error" for simplicity, so pre filtering is problematic (everyone who returned any error would have to pre filter).

----- Original Message -----
> > @@ -45,10 +45,67 @@ func (list errorList) Error() string {
> >  	return result
> >  }
> >  
> > +// Except returns any errors (or nested errors, if the list contains
> > +// nested []error slices) for which all fns return false. If no errors
> > +// remain a nil list is returned. The resulting list will have all
> > +// nested []error slices flattened as a side effect.
> 
> In the list of lists, each child list refers to a single file, right?  So
> flattening that would lose semantic information.  I'd rather see a
> FilterOut() argument that says "and go recursively" or FilterOutrecursive()
> or something.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3198/files#r22488460